File: backend/app/services/ffmpeg_builder.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from uuid import UUID

from app.models import ExportQuality, MediaMetadata, Project


@dataclass(frozen=True, slots=True)
class FfmpegPlan:
    argv: tuple[str, ...]
    filter_complex: str
    duration: float


def split_atempo(speed: float) -> tuple[float, ...]:
    if not 0.1 <= speed <= 16:
        raise ValueError("speed must be between 0.1 and 16")
    factors: list[float] = []
    remaining = speed
    while remaining > 2:
        factors.append(2.0)
        remaining /= 2
    while remaining < 0.5:
        factors.append(0.5)
        remaining /= 0.5
    factors.append(round(remaining, 6))
    return tuple(factors)


def _number(value: float) -> str:
    return f"{value:.6f}".rstrip("0").rstrip(".") or "0"


def _even(value: float) -> int:
    rounded = max(2, round(value))
    return rounded if rounded % 2 == 0 else rounded + 1


def build_ffmpeg_plan(
    project: Project,
    media: dict[UUID, tuple[MediaMetadata, Path]],
    output_path: Path,
) -> FfmpegPlan:
    clips = [
        (track, clip)
        for track in project.tracks
        for clip in track.clips
    ]
    if not clips:
        raise ValueError("no clips to export")
    for _, clip in clips:
        if clip.media_id not in media:
            raise ValueError(f"missing media: {clip.media_id}")
        metadata, _ = media[clip.media_id]
        if clip.source_end > metadata.duration + 0.05:
            raise ValueError(f"clip {clip.id} exceeds media duration")

    total_duration = max(
        clip.timeline_start + clip.output_duration for _, clip in clips
    )
    media_ids = list(dict.fromkeys(clip.media_id for _, clip in clips))
    blur_id = project.background.blur_source_id
    if blur_id is not None and blur_id not in media_ids:
        if blur_id not in media:
            raise ValueError(f"missing blur media: {blur_id}")
        media_ids.append(blur_id)
    input_index = {media_id: index for index, media_id in enumerate(media_ids)}
    inputs: list[tuple[str, str]] = [
        ("file", str(media[media_id][1])) for media_id in media_ids
    ]

    filters: list[str] = []
    labels = 0

    def label(prefix: str) -> str:
        nonlocal labels
        labels += 1
        return f"{prefix}{labels}"

    width, height = project.width, project.height
    background = project.background
    color1 = background.color.removeprefix("#")
    duration = _number(total_duration)
    if background.type == "gradient":
        color2 = background.color2.removeprefix("#")
        filters.extend(
            (
                f"color=c=0x{color1}:s={width}x{height}:d={duration}[g1]",
                f"color=c=0x{color2}:s={width}x{height}:d={duration}[g2]",
                "[g1][g2]blend=all_expr='A*(1-Y/H)+B*(Y/H)'[bg]",
            )
        )
    elif background.type == "blur":
        assert blur_id is not None
        blur_index = input_index[blur_id]
        filters.append(
            f"[{blur_index}:v]scale={width}:{height}:"
            "force_original_aspect_ratio=increase,"
            f"crop={width}:{height},boxblur={_number(background.blur_radius)}:5,"
            f"setsar=1,tpad=stop_mode=clone:stop_duration={duration}[bg]"
        )
    else:
        filters.append(
            f"color=c=0x{color1}:s={width}x{height}:d={duration}[bg]"
        )

    visual: list[tuple[int, float, str, int, int, float, float]] = []
    audio_labels: list[str] = []
    for track_order, (track, clip) in enumerate(clips):
        index = input_index[clip.media_id]
        start = _number(clip.source_start)
        end = _number(clip.source_end)
        timeline_start = _number(clip.timeline_start)

        if track.type != "audio":
            current = label("trim")
            filters.append(
                f"[{index}:v]trim=start={start}:end={end},"
                f"setpts=PTS-STARTPTS[{current}]"
            )

            crop = clip.crop
            if crop.x or crop.y or crop.w < 1 or crop.h < 1:
                cropped = label("crop")
                filters.append(
                    f"[{current}]crop=iw*{_number(crop.w)}:ih*{_number(crop.h)}:"
                    f"iw*{_number(crop.x)}:ih*{_number(crop.y)}[{cropped}]"
                )
                current = cropped

            operations: list[str] = []
            transform = clip.transform
            if transform.flip_x:
                operations.append("hflip")
            if transform.flip_y:
                operations.append("vflip")
            if transform.rotation == 90:
                operations.append("transpose=1")
            elif transform.rotation == 180:
                operations.extend(("hflip", "vflip"))
            elif transform.rotation == 270:
                operations.append("transpose=2")
            if operations:
                transformed = label("orient")
                filters.append(f"[{current}]{','.join(operations)}[{transformed}]")
                current = transformed

            if clip.speed != 1:
                sped = label("speed")
                filters.append(
                    f"[{current}]setpts=PTS/{_number(clip.speed)}[{sped}]"
                )
                current = sped

            target_width = _even(transform.width * width * transform.scale)
            target_height = _even(transform.height * height * transform.scale)
            scaled = label("scale")
            filters.append(
                f"[{current}]scale={target_width}:{target_height},"
                f"setsar=1,format=rgba[{scaled}]"
            )
            positioned = label("position")
            filters.append(
                f"[{scaled}]setpts=PTS-STARTPTS+{timeline_start}/TB[{positioned}]"
            )
            visual.append(
                (
                    1 if track.type == "overlay" else 0,
                    clip.timeline_start,
                    positioned,
                    round(transform.x * width),
                    round(transform.y * height),
                    clip.timeline_start,
                    clip.timeline_start + clip.output_duration,
                )
            )

        metadata, _ = media[clip.media_id]
        audio = clip.audio
        if (
            metadata.has_audio
            and not track.muted
            and audio.enabled
            and not audio.mute
            and audio.volume > 0
        ):
            current_audio = label("atrim")
            filters.append(
                f"[{index}:a]atrim=start={start}:end={end},"
                f"asetpts=PTS-STARTPTS,aresample=48000[{current_audio}]"
            )
            if clip.speed != 1:
                for factor in split_atempo(clip.speed):
                    adjusted = label("atempo")
                    filters.append(
                        f"[{current_audio}]atempo={_number(factor)}[{adjusted}]"
                    )
                    current_audio = adjusted
            if audio.volume != 1:
                adjusted = label("volume")
                filters.append(
                    f"[{current_audio}]volume={_number(audio.volume)}[{adjusted}]"
                )
                current_audio = adjusted
            delayed = label("delay")
            delay_ms = round(clip.timeline_start * 1000)
            filters.append(
                f"[{current_audio}]adelay={delay_ms}|{delay_ms}[{delayed}]"
            )
            audio_labels.append(delayed)

    visual.sort(key=lambda item: (item[0], item[1]))
    base = "bg"
    for index, (_, _, video_label, x, y, start, end) in enumerate(visual):
        result = "vcomposed" if index == len(visual) - 1 else label("overlay")
        filters.append(
            f"[{base}][{video_label}]overlay=x={x}:y={y}:eof_action=pass:"
            f"enable='between(t\\,{_number(start)}\\,{_number(end)})'[{result}]"
        )
        base = result
    filters.append(
        f"[{base}]trim=duration={duration},setpts=PTS-STARTPTS,"
        f"fps={project.export.fps},format=yuv420p[vout]"
    )

    if audio_labels:
        joined = "".join(f"[{item}]" for item in audio_labels)
        if len(audio_labels) == 1:
            filters.append(
                f"{joined}atrim=duration={duration},asetpts=PTS-STARTPTS[aout]"
            )
        else:
            filters.append(
                f"{joined}amix=inputs={len(audio_labels)}:duration=longest:"
                "dropout_transition=0,"
                f"atrim=duration={duration},asetpts=PTS-STARTPTS[aout]"
            )
    else:
        silence_index = len(inputs)
        inputs.append(
            ("lavfi", "anullsrc=channel_layout=stereo:sample_rate=48000")
        )
        filters.append(
            f"[{silence_index}:a]atrim=duration={duration},"
            "asetpts=PTS-STARTPTS[aout]"
        )

    quality = {
        ExportQuality.draft: ("veryfast", "28"),
        ExportQuality.standard: ("medium", "23"),
        ExportQuality.high: ("slow", "18"),
    }[project.export.quality]
    argv: list[str] = ["-hide_banner", "-nostdin"]
    for kind, source in inputs:
        if kind == "lavfi":
            argv.extend(("-f", "lavfi", "-i", source))
        else:
            argv.extend(("-i", source))
    filter_complex = ";".join(filters)
    argv.extend(
        (
            "-filter_complex",
            filter_complex,
            "-map",
            "[vout]",
            "-map",
            "[aout]",
            "-c:v",
            project.export.video_codec,
            "-preset",
            quality[0],
            "-crf",
            quality[1],
            "-pix_fmt",
            "yuv420p",
            "-c:a",
            "aac",
            "-b:a",
            project.export.audio_bitrate,
            "-movflags",
            "+faststart",
            "-t",
            duration,
            "-progress",
            "pipe:1",
            "-nostats",
            "-y",
            str(output_path),
        )
    )
    return FfmpegPlan(tuple(argv), filter_complex, total_duration)

File: server/services/ffmpeg_builder.js
function splitAtempo(speed) {
  if (speed < 0.1 || speed > 16) throw new Error("Speed must be between 0.1 and 16");
  const factors = [];
  let remaining = speed;
  while (remaining > 2) {
    factors.push(2.0);
    remaining /= 2;
  }
  while (remaining < 0.5) {
    factors.push(0.5);
    remaining /= 0.5;
  }
  factors.push(Math.round(remaining * 1000000) / 1000000);
  return factors;
}

function numStr(val) {
  const num = Number(val ?? 0);
  return (isNaN(num) ? 0 : num).toFixed(6);
}

function evenInt(val) {
  const rounded = Math.max(2, Math.round(val ?? 2));
  return rounded % 2 === 0 ? rounded : rounded + 1;
}

function buildFfmpegPlan(project, resolvedMediaMap, outputPath) {
  const clips = [];
  for (const track of project.tracks || []) {
    for (const clip of track.clips || []) {
      clips.push({ track, clip });
    }
  }

  if (clips.length === 0) {
    throw new Error("No clips to export");
  }

  let maxDuration = 0;
  for (const { clip } of clips) {
    const sStart = clip.sourceStart ?? clip.source_start ?? 0;
    const sEnd = clip.sourceEnd ?? clip.source_end ?? 0;
    const tStart = clip.timelineStart ?? clip.timeline_start ?? 0;
    const speed = clip.speed || 1;
    const outDur = (sEnd - sStart) / speed;
    const endPos = tStart + outDur;
    if (endPos > maxDuration) maxDuration = endPos;
  }

  const mediaIds = Array.from(new Set(clips.map(c => c.clip.media_id || c.clip.mediaId)));
  const blurId = project.background?.blur_source_id || project.background?.blurSourceId;
  if (blurId && !mediaIds.includes(blurId)) {
    mediaIds.push(blurId);
  }

  const inputIndexMap = {};
  mediaIds.forEach((id, idx) => { inputIndexMap[id] = idx; });

  const inputs = mediaIds.map(id => {
    const mediaObj = resolvedMediaMap[id];
    if (!mediaObj) throw new Error(`Missing media file for ID: ${id}`);
    if (mediaObj.metadata?.is_image) {
      // 이미지는 -loop 1로 무한 반복 후 타임라인 길이만큼 자름
      return ['-loop', '1', '-framerate', '30', '-i', mediaObj.path];
    }
    return ['-i', mediaObj.path];
  });

  const filters = [];
  let labelCounter = 0;
  function nextLabel(prefix) {
    labelCounter++;
    return `${prefix}${labelCounter}`;
  }

  const exportSettings = project.export || {};
  const width = exportSettings.width || 1920;
  const height = exportSettings.height || 1080;
  const totalDurationStr = numStr(maxDuration);
  const bg = project.background || { type: 'solid', color: '#000000' };
  const color1 = (bg.color || '#000000').replace('#', '');

  if (bg.type === 'gradient') {
    const color2 = (bg.color2 || '#1a1a2e').replace('#', '');
    filters.push(
      `color=c=0x${color1}:s=${width}x${height}:d=${totalDurationStr}[g1]`,
      `color=c=0x${color2}:s=${width}x${height}:d=${totalDurationStr}[g2]`,
      `[g1][g2]blend=all_expr='A*(1-Y/H)+B*(Y/H)'[bg]`
    );
  } else if (bg.type === 'blur' && blurId) {
    const bIdx = inputIndexMap[blurId];
    const radius = bg.blur_radius || bg.blur || 20;
    filters.push(
      `[${bIdx}:v]scale=${width}:${height}:force_original_aspect_ratio=increase,crop=${width}:${height},boxblur=${numStr(radius)}:5,setsar=1,tpad=stop_mode=clone:stop_duration=${totalDurationStr}[bg]`
    );
  } else {
    filters.push(`color=c=0x${color1}:s=${width}x${height}:d=${totalDurationStr}[bg]`);
  }

  const visualOverlays = [];
  const audioLabels = [];

  for (const { track, clip } of clips) {
    const mId = clip.media_id || clip.mediaId;
    const idx = inputIndexMap[mId];
    const sStart = clip.sourceStart ?? clip.source_start ?? 0;
    const sEnd = clip.sourceEnd ?? clip.source_end ?? 0;
    const tStart = clip.timelineStart ?? clip.timeline_start ?? 0;
    const speed = clip.speed || 1;
    const startStr = numStr(sStart);
    const endStr = numStr(sEnd);
    const tStartStr = numStr(tStart);
    const outDur = (sEnd - sStart) / speed;

    if (track.type !== 'audio' && track.kind !== 'audio') {
      let curr = nextLabel('trim');
      filters.push(`[${idx}:v]trim=start=${startStr}:end=${endStr},setpts=PTS-STARTPTS[${curr}]`);

      const crop = clip.crop || {};
      if (crop.x || crop.y || (crop.w && crop.w < 1) || (crop.h && crop.h < 1)) {
        const cw = crop.w || crop.width || 1;
        const ch = crop.h || crop.height || 1;
        const cx = crop.x || 0;
        const cy = crop.y || 0;
        const cropped = nextLabel('crop');
        filters.push(`[${curr}]crop=iw*${numStr(cw)}:ih*${numStr(ch)}:iw*${numStr(cx)}:ih*${numStr(cy)}[${cropped}]`);
        curr = cropped;
      }

      const tf = clip.transform || {};
      const ops = [];
      if (tf.flip_x || tf.flipX) ops.push('hflip');
      if (tf.flip_y || tf.flipY) ops.push('vflip');
      if (tf.rotation === 90) ops.push('transpose=1');
      else if (tf.rotation === 180) ops.push('hflip,vflip');
      else if (tf.rotation === 270) ops.push('transpose=2');

      if (ops.length > 0) {
        const oriented = nextLabel('orient');
        filters.push(`[${curr}]${ops.join(',')}[${oriented}]`);
        curr = oriented;
      }

      if (speed !== 1) {
        const sped = nextLabel('speed');
        filters.push(`[${curr}]setpts=PTS/${numStr(speed)}[${sped}]`);
        curr = sped;
      }

      const scale = tf.scale || 1;
      const tfW = tf.width || 1;
      const tfH = tf.height || 1;
      const targetW = evenInt(tfW * width * scale);
      const targetH = evenInt(tfH * height * scale);

      const scaled = nextLabel('scale');
      filters.push(`[${curr}]scale=${targetW}:${targetH},setsar=1,format=rgba[${scaled}]`);
      curr = scaled;

      const pos = nextLabel('position');
      filters.push(`[${curr}]setpts=PTS-STARTPTS+${tStartStr}/TB[${pos}]`);

      const posX = Math.round((tf.x || 0) * width);
      const posY = Math.round((tf.y || 0) * height);
      const tStart = clip.timeline_start || 0;
      visualOverlays.push({
        isOverlay: (track.type === 'overlay' || track.kind === 'overlay') ? 1 : 0,
        timelineStart: tStart,
        label: pos,
        x: posX,
        y: posY,
        start: tStart,
        end: tStart + outDur
      });
    }

    const metaObj = resolvedMediaMap[mId]?.metadata || {};
    const audio = clip.audio || { enabled: true, volume: 1, mute: false };
    if (metaObj.has_audio && !track.muted && audio.enabled && !audio.mute && (audio.volume || 1) > 0) {
      let currA = nextLabel('atrim');
      filters.push(`[${idx}:a]atrim=start=${startStr}:end=${endStr},asetpts=PTS-STARTPTS,aresample=48000[${currA}]`);

      if (speed !== 1) {
        const factors = splitAtempo(speed);
        for (const factor of factors) {
          const adj = nextLabel('atempo');
          filters.push(`[${currA}]atempo=${numStr(factor)}[${adj}]`);
          currA = adj;
        }
      }

      const vol = audio.volume !== undefined ? audio.volume : 1;
      if (vol !== 1) {
        const volLabel = nextLabel('volume');
        filters.push(`[${currA}]volume=${numStr(vol)}[${volLabel}]`);
        currA = volLabel;
      }

      const delayMs = Math.round((clip.timeline_start || 0) * 1000);
      const delayed = nextLabel('delay');
      filters.push(`[${currA}]adelay=${delayMs}|${delayMs}[${delayed}]`);
      audioLabels.push(delayed);
    }
  }

  visualOverlays.sort((a, b) => {
    if (a.isOverlay !== b.isOverlay) return a.isOverlay - b.isOverlay;
    return a.timelineStart - b.timelineStart;
  });

  let base = 'bg';
  visualOverlays.forEach((vo, idx) => {
    const isLast = idx === visualOverlays.length - 1;
    const resLabel = isLast ? 'vcomposed' : nextLabel('overlay');
    filters.push(
      `[${base}][${vo.label}]overlay=x=${vo.x}:y=${vo.y}:eof_action=pass:enable='between(t\\,${numStr(vo.start)}\\,${numStr(vo.end)})'[${resLabel}]`
    );
    base = resLabel;
  });

  const fps = exportSettings.fps || 30;
  filters.push(`[${base}]trim=duration=${totalDurationStr},setpts=PTS-STARTPTS,fps=${fps},format=yuv420p[vout]`);

  let hasExtraAudioInput = false;
  if (audioLabels.length === 1) {
    filters.push(`[${audioLabels[0]}]atrim=duration=${totalDurationStr},asetpts=PTS-STARTPTS[aout]`);
  } else if (audioLabels.length > 1) {
    const joined = audioLabels.map(l => `[${l}]`).join('');
    filters.push(`${joined}amix=inputs=${audioLabels.length}:duration=longest:dropout_transition=0,atrim=duration=${totalDurationStr},asetpts=PTS-STARTPTS[aout]`);
  } else {
    hasExtraAudioInput = true;
    inputs.push(['-f', 'lavfi', '-i', 'anullsrc=channel_layout=stereo:sample_rate=48000']);
    const nullIdx = inputs.length - 1;
    filters.push(`[${nullIdx}:a]atrim=duration=${totalDurationStr},asetpts=PTS-STARTPTS[aout]`);
  }

  const qualityPreset = {
    draft: ['veryfast', '28'],
    standard: ['medium', '23'],
    high: ['slow', '18']
  }[exportSettings.quality || 'standard'] || ['medium', '23'];

  const argv = ['-hide_banner', '-nostdin'];
  inputs.forEach(inp => argv.push(...inp));

  argv.push(
    '-filter_complex', filters.join(';'),
    '-map', '[vout]',
    '-map', '[aout]',
    '-c:v', exportSettings.video_codec || 'libx264',
    '-preset', qualityPreset[0],
    '-crf', qualityPreset[1],
    '-pix_fmt', 'yuv420p',
    '-c:a', 'aac',
    '-b:a', exportSettings.audio_bitrate || '192k',
    '-movflags', '+faststart',
    '-t', totalDurationStr,
    '-progress', 'pipe:1',
    '-nostats',
    '-y',
    outputPath
  );

  return { argv, duration: maxDuration };
}

module.exports = { buildFfmpegPlan };
